### DISCUSSION TITLE
Guard against #including new and old version in the same file

### DISCUSSION THREAD

**Discussion by mortenfyhn** [nlohmann/json#3337] [Ideas]
I recently had a case where we included this library (v3.9.1) and unexpectedly got a bunch of redefinition errors. After some digging, it turned out we indirectly included a much older version of this library (v2.1.1), which was being used internally in some header. These 
versions use different header guards, so the double inclusion wasn't avoided that way.

The old guard is `NLOHMANN_JSON_HPP` and the new one is `INCLUDE_NLOHMANN_JSON_HPP_`.

What about checking for `#ifdef NLOHMANN_JSON_HPP`, and if true, show a warning (with `#warning` maybe)? That way the developer will quickly realize what's going on.

I suppose that will only help if you include the new version after the old version, but it seems better than nothing.

**Comment by nlohmann** [comment]
As the include guard is undefined at the end of the `json.hpp` header, I am not sure how this would fix the issue.

**Reply by gregmarr** [reply]
Why would the include guard be undefined at the end?  That would make it useless, unless it's only intended for preventing duplicate includes within the multi-file includes, and not preventing a user from including it multiple times.

**Reply by nlohmann** [reply]
Stupid me - sorry for the mistake. Yes, I guess then we can make a check for this - but I guess then it would make more sense to check for NLOHMANN_JSON_VERSION_MAJOR, NLOHMANN_JSON_VERSION_MINOR, and NLOHMANN_JSON_VERSION_PATCH. What do you think?

**Reply by gregmarr** [reply]
Yeah, that could be useful.

**Reply by nlohmann** [reply]
@mortenfyhn @gregmarr Please check https://github.com/nlohmann/json/pull/3418.
